### ISSUE TITLE
Flaky CI: blur-(5,5) norm-1 test returns objective 9.32 instead of 6.96 on macOS

### ISSUE THREAD

**Issue by vtjeng** [vtjeng/MIPVerify.jl#219]
The `target labels = 2, blur-(5,5) perturbation, norm order = 1` case in `test/integration/sequential/generated_weights/conv+fc+softmax.jl` failed on the Julia 1 (resolved 1.12.6) / macos-latest / x64 CI job with a much worse objective value than expected:

```
Expression: isapprox(actual_objective_value, expected_objective_value; rtol = 5.0e-5)
 Evaluated: isapprox(9.323264824200079, 6.959316; rtol = 5.0e-5)
```

- Job: https://github.com/vtjeng/MIPVerify.jl/actions/runs/29267847943/job/86877738113
- Commit: d4b9084 (PR #214)
- Versions: Julia 1.12.6, HiGHS.jl v1.24.1, HiGHS_jll v1.15.1+0
- The same case passed on the other five platform jobs in the same run, and on master with the same dependency versions the same day ([run 29233189449](https://github.com/vtjeng/MIPVerify.jl/actions/runs/29233189449)).

Diagnostic notes:

- `test_find_adversarial_example` (test/TestHelpers.jl:87) checks the objective value without first asserting `d[:SolveStatus] == OPTIMAL`, so the log cannot distinguish between HiGHS claiming optimality at 9.32 (a solver wrong answer) and early termination with a feasible incumbent. Asserting the solve status before the objective comparison would make future occurrences diagnosable.
- The main solve has no time limit, but tightening solves are capped at 20 s (`get_tightening_options`). On a slow runner, tightening timeouts fall back to looser interval-arithmetic bounds; the formulation stays valid and the true optimum should not change, but the larger big-M coefficients could affect solver numerics.

## Occurrences

| Date       | Branch/PR      | Job                                                                                |
| ---------- | -------------- | ---------------------------------------------------------------------------------- |
| 2026-07-13 | #214 (d4b9084) | [Julia 1 - macos-latest - x64](https://github.com/vtjeng/MIPVerify.jl/actions/runs/29267847943/job/86877738113) |
| 2026-07-16 | #230 (5a0756c) | [Julia 1.12 - macos-latest - x64](https://github.com/vtjeng/MIPVerify.jl/actions/runs/29532850660/job/87737117872) |
| 2026-07-16 | master (a6473c7, merge of #236) | [Julia 1 - macos-latest - x64](https://github.com/vtjeng/MIPVerify.jl/actions/runs/29536865449/job/87750161989) |

Append new occurrences to this table.

**Comment by vtjeng** [comment]
First data point after filing: the next CI run on the same PR ([007eda8](https://github.com/vtjeng/MIPVerify.jl/pull/214), same HiGHS.jl v1.24.1 / Julia 1.12.6) passed the Julia 1 / macOS job with the expected objective, so the mismatch did not reproduce on the immediately following run.

**Comment by vtjeng** [comment]
Investigated; root cause is HiGHS 1.15.1 itself (same as #218), and #220 pins the test environment to HiGHS_jll 1.14 as a mitigation.

**Diagnosis: the parallel MIP solver introduced in HiGHS 1.15 returned a wrong answer.** HiGHS 1.15.0 (2026-06-28) shipped the first variant of the parallel MIP solver; CI picked up jll 1.15.1+0 unpinned in early July, and within two days of exposure CI produced three HiGHS failures on three platforms: this issue's wrong objective (macOS), and the #218 segfaults (ubuntu 2026-07-13, windows 2026-07-14).

Evidence gathered locally (Linux x64, Julia 1.12.6, HiGHS.jl v1.24.1, jll 1.15.1+0, same artifact hash as the failing CI jobs):

- Repeatedly re-solving this exact test's MIP crashed HiGHS at two distinct sites in the new MIP machinery: a bus error in `HighsSplitDeque::WorkerBunk::publishWork` (parallel root-node heuristics), and a segfault in `HighsPrimalHeuristics::RINS → solveSubMip → HighsProfiling::running`. The second crash occurred with `threads=1, parallel=off`, so solver options do not avoid the new code. The #218 ubuntu stack (`HighsMipSolver::run → HighsSeparation → HighsLpRelaxation → HighsProfiling::isSubMip`) is a third distinct site in the same machinery.
- The crash-site symbols (`HighsMipWorker`, `HighsProfiling`) do not exist in jll 1.14.0's `libhighs.so` (verified with `nm`); the crashing code paths are new in 1.15.
- The formulation is exonerated. With pure interval-arithmetic bounds (the loosest valid formulation, which is what tightening-solve timeouts on a slow runner would produce), the MIP still solves to 6.959316 OPTIMAL every time. The LP-certificate bound math is valid independent of solver duals, so platform BLAS differences (macOS uses Apple Accelerate) can only loosen bounds, never cut off the optimum.
- Solver-independent proof that 9.32-as-optimal is wrong: at the solution found locally, `primal_feasibility_report` shows a worst constraint violation of 8.9e-16, the norm-1 objective recomputed by hand is 6.959315922990173, and a pure-Julia forward pass of the perturbed input ties the target logit with the runner-up. A feasible point at 6.96 exists, so any claim that 9.32 is optimal is a solver wrong answer.
- Intermittency is consistent: the passing master run 29233189449 also used jll 1.15.1+0 (verified from its log), and one local stress trial survived 40 re-solves. HiGHS 1.15.1 is also nondeterministic run to run on identical inputs (tightening-LP duals vary by ~1e-12, final MIP dual bounds differ across identical runs).

Upstream reports consistent with this: ERGO-Code/HiGHS#3150 (1.15.1 crash in parallel MIP, filed 2026-07-15), ERGO-Code/HiGHS#2885.

A before/after stress comparison (4 trials x 40 re-solves per jll version) is running; I will post the tally here. #220 also adds the solve-status assertion suggested in this issue's diagnostic notes, so any future occurrence is unambiguous in the log.

**Comment by vtjeng** [comment]
Another occurrence, 2026-07-16, on PR #209 CI (Julia 1.12.6, macos-latest): `isapprox(7.3603039035631985, 6.959316; rtol = 5.0e-5)` failed in the same `target labels = 2, blur-(5,5) perturbation, norm order = 1` test. Job: https://github.com/vtjeng/MIPVerify.jl/actions/runs/29485396706/job/87578618293 — third distinct objective value seen for this test on macOS.

**Comment by vtjeng** [comment]
Stress-test tally, as promised. Correction to the comment above first: there were three local crashes on jll 1.15.1, not two (the RINS/HighsProfiling segfault reproduced on an immediate rerun of the same script, 2-for-2).

Counted measurements on this machine (Linux x64, Julia 1.12.6), each cell = processes without a crash / total processes:

| Workload shape | jll 1.15.1 | jll 1.14.0 |
| --- | --- | --- |
| flip re-solve sweeps (200/process, 880 solves total) | 6/6 | 6/6 |
| 4 time-limit-truncated solves then full solves | 4/4 | 4/4 |
| full 521-test suite in one process | 3/3 | 3/3 |
| exploratory probes (first session hour) | 1/4 (3 crashes) | n/a |

Interpretation: the 1.15.1 crashes are real (three full stacks in the new MIP-worker/profiling code, symbols absent from 1.14.0) but layout-sensitive and low-probability per process on this hardware; counted reproductions of the exact crashing shapes all survived. This is characteristic of heap corruption, where crashing versus silently returning a wrong result is allocation luck. Crash rate is a lower bound on the corruption rate; the 9.32 objective is the silent manifestation.

CI context: 3 failures in roughly 24 test-job executions on 1.15.1 over 2026-07-13/14 (about 12% per job, 3 platforms). A 0-for-3 local suite result is consistent with that rate (P ≈ 0.67), and the 2-core contended CI runners plausibly trigger the races more often than a idle workstation.

Suite runtime is unaffected by the pin: mean 4m29s (1.15.1) vs 4m18s (1.14.0) over 3 paired runs each.

Earlier wording ("segfaults readily") reflected the exploratory hour, not the measured rate; PR #220's description has been updated accordingly.

**Comment by vtjeng** [comment]
**Update: the segfault now reproduces locally on demand, and it is pinned to the 1.15-only parallel-MIP code.**

Reproducer: a standalone script that builds the same MIP as the failing test case (`conv+fc+softmax` generated weights, target label 2, blur-(5,5) perturbation, norm order 1) and solves it repeatedly in one process:

1. 4 solves with the main MIP truncated at `time_limit = 0.01` (LP bound tightening runs to completion each time);
2. 4 full solves with `threads = 1`, `parallel = "off"`;
3. 4 full solves with default options.

**On HiGHS_jll 1.15.1+0** (linux x86_64, Julia 1.12.6, HiGHS.jl v1.24.1): **3 segfaults in 5 runs** of this script — the two initial runs crashed back-to-back, and 1 of 3 controlled reruns crashed. Every crash:

- happened in phase 2, i.e. with **`threads = 1`, `parallel = "off"`** — single-threaded options do not avoid the new code — on the 2nd, 2nd, and 4th full solve respectively;
- has the same stack: `HighsPrimalHeuristics::RINS(HighsMipWorker&, …)` → `HighsSearch::evaluateNode` → `HighsLpRelaxation::resolveLp/run`, faulting inside `HighsProfiling` (`::running` twice, `::isSubMip` once). `HighsMipWorker` and `HighsProfiling` exist only in the 1.15.x artifact; `nm -D` on 1.14.0's `libhighs.so` shows none of this machinery.

**On HiGHS_jll 1.14.0+0** (same machine, same script, HiGHS.jl v1.23.0): **0 segfaults in 3 runs**, every full solve `OPTIMAL` at 6.959316.

Two observations from the non-crashing runs:

- Tightened variable bounds differ run-to-run only at the ~1e-13 level, on both versions — ordinary LP-dual nondeterminism, harmless here because the formulation's bounds are certified independently of solver duals.
- The crash is site-deterministic but not run-deterministic: every crash lands at the same RINS/profiling site, yet only ~half of runs crash. That is the signature of heap corruption whose visibility depends on allocator layout — the wild write usually lands on mapped memory and corrupts silently. It supports reading the original failure here (a wrong objective of 9.323… reported as `OPTIMAL`) as a corrupted incumbent rather than a formulation or tolerance problem, and it means the observed crash rate is a lower bound on the corruption rate.

The CI failures were on macOS (#219) and ubuntu/windows (#218); this reproduces on linux, so the problem is not platform-specific.

<details>
<summary>Crash stack (controlled rerun, rep 1)</summary>

```
[715711] signal 11 (1): Segmentation fault
in expression starting at probe3_219.jl:94
_ZN14HighsProfiling8isSubMipEv at ~/.julia/artifacts/1817ed3230ae7addaac98302147c1b6bf84d8b33/lib/libhighs.so
_ZN17HighsLpRelaxation3runEb at libhighs.so
_ZN17HighsLpRelaxation9resolveLpEP11HighsDomain at libhighs.so
_ZN11HighsSearch12evaluateNodeEv at libhighs.so
_ZN21HighsPrimalHeuristics4RINSER14HighsMipWorkerRKSt6vectorIdSaIdEE at libhighs.so
_ZZZN14HighsMipSolver3runEvENKUlRKSt6vectorIiSaIiEERS0_IZNS_3runEvE11RestartVoteSaIS5_EEbiidE21_clES4_S8_biidENKUliE_clEi at libhighs.so
_ZN14HighsMipSolver3runEv at libhighs.so
_ZN5Highs12callSolveMipEv at libhighs.so
_ZN5Highs19calledOptimizeModelEv at libhighs.so
_ZN5Highs13optimizeModelEv at libhighs.so
_ZN5Highs3runEv at libhighs.so
Highs_run at ~/.julia/packages/HiGHS/iY2Ae/src/gen/libhighs.jl:344 [inlined]
optimize! at ~/.julia/packages/HiGHS/iY2Ae/src/MOI_wrapper.jl:2300
...
find_adversarial_example at src/MIPVerify.jl:117
```

</details>

<details>
<summary>Reproducer script (probe3_219.jl)</summary>

Run with MIPVerify dev'd from this repo and HiGHS_jll pinned (e.g. in a scratch env: `Pkg.add(name = "HiGHS_jll", version = "1.15.1")`, then `Pkg.develop(path = <repo>)`). Set `MIPVERIFY_REPO` to the repo root so the script can include `test/TestHelpers.jl`.

```julia
# Probe 3 for issue 219: isolate the source of nondeterminism.
#  A. Tightening only (main solve truncated at 10ms): are post-tightening variable
#     bounds bit-identical across runs?
#  B. Full solve with parallel disabled (threads=1, parallel=off): deterministic?
#  C. Full solve with defaults: nondeterministic? (baseline)

using JuMP
using MathOptInterface
const MOI = MathOptInterface
using HiGHS
using MIPVerify
using MIPVerify: BlurringPerturbationFamily, find_adversarial_example, lp

const REPO = get(ENV, "MIPVERIFY_REPO", ".")
include(joinpath(REPO, "test", "TestHelpers.jl"))

function build_nn()
    input = gen_array((1, 6, 6, 1), 0, 1)
    c1_kernel = gen_array((2, 2, 1, 2), -1, 1)
    c1_bias = gen_array((2,), -1, 1)
    l1_kernel = gen_array((18, 4), -1, 1)
    l1_bias = gen_array((4,), -1, 1)
    l2_kernel = gen_array((4, 3), -1, 1)
    l2_bias = gen_array((3,), -1, 1)
    nn = Sequential(
        [
            Conv2d(c1_kernel, c1_bias),
            MaxPool((1, 2, 2, 1)),
            ReLU(),
            Flatten(4),
            Linear(l1_kernel, l1_bias),
            ReLU(),
            Linear(l2_kernel, l2_bias),
        ],
        "probe219",
    )
    return nn, input
end

function solve_case(nn, input; main_options = Dict("output_flag" => false), tightening_extra = Dict())
    topts = convert(
        Dict{String,Any},
        merge(Dict("output_flag" => false, "time_limit" => 20.0), tightening_extra),
    )
    d = find_adversarial_example(
        nn,
        input,
        2,
        HiGHS.Optimizer,
        main_options;
        pp = BlurringPerturbationFamily((5, 5)),
        norm_order = 1,
        tightening_options = topts,
        tightening_algorithm = lp,
    )
    return d
end

function bound_signature(m)
    sig = Float64[]
    for v in JuMP.all_variables(m)
        push!(sig, JuMP.has_lower_bound(v) ? JuMP.lower_bound(v) : -Inf)
        push!(sig, JuMP.has_upper_bound(v) ? JuMP.upper_bound(v) : Inf)
    end
    return sig
end

function compare_sigs(sigs)
    base = sigs[1]
    for i in 2:length(sigs)
        diffs = [k for k in eachindex(base) if !isequal(base[k], sigs[i][k])]
        if isempty(diffs)
            println("    run $i: bounds bit-identical to run 1")
        else
            mx = maximum(abs(base[k] - sigs[i][k]) for k in diffs)
            println("    run $i: $(length(diffs)) bound entries differ, max abs diff $mx")
        end
    end
end

nn, input = build_nn()

println("=== A. Tightening determinism (main solve capped at 10ms) ===")
sigsA = []
for i in 1:4
    d = solve_case(nn, input; main_options = Dict("output_flag" => false, "time_limit" => 0.01))
    push!(sigsA, bound_signature(d[:Model]))
    println("  run $i: main status=$(d[:SolveStatus])")
end
compare_sigs(sigsA)

println("=== B. Full solve, threads=1 / parallel=off ===")
sigsB = []
for i in 1:4
    d = solve_case(
        nn,
        input;
        main_options = Dict("output_flag" => false, "threads" => 1, "parallel" => "off"),
    )
    push!(sigsB, bound_signature(d[:Model]))
    m = d[:Model]
    println("  run $i: status=$(d[:SolveStatus]) obj=$(JuMP.objective_value(m)) bound=$(JuMP.objective_bound(m))")
end
compare_sigs(sigsB)

println("=== C. Full solve, default options (baseline) ===")
for i in 1:4
    d = solve_case(nn, input)
    m = d[:Model]
    println("  run $i: status=$(d[:SolveStatus]) obj=$(JuMP.objective_value(m)) bound=$(JuMP.objective_bound(m))")
end
```

</details>

**Comment by vtjeng** [comment]
Second occurrence, appended to the table: [Julia 1.12 - macos-latest - x64](https://github.com/vtjeng/MIPVerify.jl/actions/runs/29532850660/job/87737117872) on PR #230 (`5a0756c`), 2026-07-16.

Same assertion at test/TestHelpers.jl:88, expected 6.959316, actual **8.437194890835396** (first occurrence was 9.323264824200079 — a different wrong value, consistent with nondeterministic solver behavior rather than a stable wrong answer). Both occurrences are macOS-only. This one is the Julia 1.12 leg where the first was the Julia 1 leg, but both selectors currently resolve to Julia 1.12.6, so the occurrences say nothing about version sensitivity yet.

The job log shows `Running HiGHS 1.15.1`: #220, which pins HiGHS_jll below 1.15 and adds the solve-status assertion this issue's diagnostic notes call for, has not merged, so the suspect solver version was still in play. The other five legs in the run passed; rerunning the failed job.

**Comment by vtjeng** [comment]
Third occurrence, appended to the table: [Julia 1 - macos-latest - x64](https://github.com/vtjeng/MIPVerify.jl/actions/runs/29536865449/job/87750161989) on master `a6473c7` (merge of #236), 2026-07-16, Julia 1.12.6, HiGHS 1.15 still unpinned.

Same assertion, expected 6.959316, actual **7.413022603240303** — a third distinct wrong value (9.323, 8.437, 7.413), all macOS, all under Rosetta x64. This is the last master run on x64 macOS; #237 moved the macOS legs to native aarch64, so future occurrences (or their absence) will indicate whether the flake is Rosetta-related. Rerunning the failed job.
